service/controller/v3/resource/instance/create.go

```diff
@@ -33,187 +31,205 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+		// and instruct the worker instance to be reimaged again.
+		var workerUpgraded bool
+		if !masterUpgraded {
+			allWorkerInstances, err := r.allInstances(ctx, customObject, key.WorkerVMSSName)
+			if IsScaleSetNotFound(err) {
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find the scale set '%s'", key.WorkerVMSSName(customObject)))
+			} else if err != nil {
 				return microerror.Mask(err)
-			}
+			} else {
+				r.logger.LogCtx(ctx, "level", "debug", "message", "processing worker VMSSs")
 
-			r.logger.LogCtx(ctx, "level", "debug", "message", "processed worker VMSSs")
-		}
-	} else {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "not processing worker VMSSs due to master VMSSs processing")
-	}
+				workerInstanceToUpdate, workerInstanceToDrain, workerInstanceToReimage, err := r.nextInstance(ctx, customObject, allWorkerInstances, nodeConfigs, key.WorkerInstanceName, versionValue)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+				err = r.updateInstance(ctx, customObject, workerInstanceToUpdate, key.WorkerVMSSName, key.WorkerInstanceName)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+				err = r.createDrainerConfig(ctx, customObject, workerInstanceToDrain, key.WorkerInstanceName)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+				err = r.reimageInstance(ctx, customObject, workerInstanceToReimage, key.WorkerVMSSName, key.WorkerInstanceName)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+				err = r.deleteDrainerConfig(ctx, customObject, workerInstanceToReimage, key.WorkerInstanceName, nodeConfigs)
+				if err != nil {
+					return microerror.Mask(err)
+				}
 
-	{
-		r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring deployment")
+				if workerInstanceToUpdate != nil || workerInstanceToDrain != nil || workerInstanceToReimage != nil {
+					workerUpgraded = true
+				}
 
-		computedDeployment, err := r.newDeployment(ctx, customObject, nil)
-		if controllercontext.IsInvalidContext(err) {
-			r.logger.LogCtx(ctx, "level", "debug", "message", "missing dispatched output values in controller context")
-			r.logger.LogCtx(ctx, "level", "debug", "message", "not ensuring deployment")
-		} else if err != nil {
-			return microerror.Mask(err)
+				r.logger.LogCtx(ctx, "level", "debug", "message", "processed worker VMSSs")
+			}
 		} else {
-			_, err = deploymentsClient.CreateOrUpdate(ctx, key.ClusterID(customObject), vmssDeploymentName, computedDeployment)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "not processing worker VMSSs due to master VMSSs processing")
+		}
+
+		if !masterUpgraded && !workerUpgraded {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "neither masters nor workers upgraded")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "removing resource status 'Stage/InstancesUpgrading'")
+
+			err := r.deleteResourceStatus(customObject, "Stage", "InstancesUpgrading")
```

Comment:
The cycle ends here and with removal of the last stage status the next cycle starts with the next resync period.